File: baekjoon/exhaustiveSearch/1963.py
import sys
from collections import deque
input = sys.stdin.readline

primeNumList = [0] * 10001

for i in range(2, 5000):
    if primeNumList[i] == False:
        start = i
        while start + i < len(primeNumList):
            start += i
            primeNumList[start] = True

def bfs():
    q = deque([(startNum, 0)])
    while q:
        num, value = q.popleft()

        if num == endNum:
            return value

        strNum = str(num)

        value += 1
        for i in range(4):
            for j in map(str,range(10)):
                if i == 0 and j == '0':
                    continue
                nextNum = int(strNum[:i] + j + strNum[i+1:])
                if primeNumList[nextNum] == False and isVisited[nextNum] == False:
                    isVisited[nextNum] = True
                    q.append((nextNum, value))



t = int(input())

for _ in range(t):
    startNum, endNum = map(int,input().split())
    isVisited = [False] * 10001
    print(bfs())

File: baekjoon_js/exhaustiveSearch/1963.js
("use strict");

const fs = require("fs");

const stdin = (process.platform === "linux"
  ? fs.readFileSync("/dev/stdin").toString()
  : `3
1033 8179
1373 8017
1033 1033
`
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

let T = Number(input());
let primeNumList = Array(10001).fill(false);

primeNumList[0] = primeNumList[1] = true;

for (let i = 2; i < 5001; i++) {
  for (let j = i + i; j < 10001; j += i) {
    if (primeNumList[j] === false) primeNumList[j] = true;
  }
}

while (T--) {
  let [num, target] = input().split(" ").map(Number);
  let isVisited = Array(10001).fill(false);
  console.log(BFS(num, target, isVisited));
}

// let i = "1923";
// console.log(i.slice(0, 2) + i.slice(2, 4));
// let j = String(i);
// j = j.replace(j[0], "2");

// console.log(j);

function BFS(start, target, isVisited) {
  q = [];
  q.push([start, 0]);

  let nextNum;
  while (q.length > 0) {
    let [nowNum, nowCnt] = q.shift();
    if (nowNum === target) return nowCnt;
    nowCnt++;

    strNum = String(nowNum);
    for (let j = 0; j < 4; j++) {
      for (let i = 0; i < 10; i++) {
        if (j == 0 && i == 0) continue;
        // nextNum = String(nowNum);
        // nextNum = nextNum.replace(nextNum[j], String(i));

        nextNum = Number(
          strNum.slice(0, j) + String(i) + strNum.slice(j + 1, 4)
        );
        // console.log(nextNum);
        if (isVisited[nextNum] === false && primeNumList[nextNum] === false) {
          isVisited[nextNum] = true;
          q.push([nextNum, nowCnt]);
        }
      }
    }
  }
  return "Impossible";
}
